📖  CAEP to add support for managed external etcd clusters in CAPI (#4659)
kubernetes-sigs/cluster-api: docs/proposals/20210524-etcdadm-bootstrap-provider.md

- Introduce pluggable etcd providers in Cluster API, starting with an etcdadm based provider. This etcdadm provider will create and manage an etcd cluster.
- User should be able to create a Kubernetes cluster that uses external etcd topology in a single step.
- User should only have to create the required CAPI resources, along with the required etcd provider specific resources. That should trigger creation of an etcd cluster, followed by creation of the target workload cluster which uses this etcd cluster. Meaning, the CAPI controllers will deploy the etcd cluster first and then use that as input to the workload cluster, without requiring any user intervention.
- There will be a 1:1 mapping between the external etcd cluster and the workload cluster.
- Define a contract for pluggable etcd providers and define steps that control plane providers should take to use a managed external etcd cluster.
- Support following etcd cluster management actions: scale up and scale down, etcd member replacement, etcd version upgrades and rollbacks.
- The etcd providers will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.

> ### Non-Goals

Author: per our discussion, added it as a non-goal for this iteration

- Then the controller will first initialize etcd cluster by creating one EtcdadmConfig and a corresponding Machine resource. 
- The CAPI [machine controller](https://github.com/mrajashree/cluster-api/blob/etcdadm_bootstrap/controllers/machine_controller_phases.go#L322-L378) gets the machine's IP address/hostname from the InfraMachine, and as soon as the first etcd machine's IP is obtained, it will create a Secret to store this address, and set the `Initialized` condition on the EtcdCluster object. This step is required because etcdadm join command must have the first node's client URL.
- Once the EtcdCluster is Initialized, the etcd cluster controller will scale up the cluster to add as many members as required to match EtcdCluster.Spec.Replicas field
- At the end of every reconciliation loop, the etcd cluster controller will check if desired number of members are created by
    * Getting Machines owned by EtcdCluster
    * Running healthcheck for each Machine with the etcd `/health` endpoint at `https://client-url:2379/health`

> - Once all members are ready, EtcdCluster will set the endpoints on EtcdCluster.Status.Endpoints field, and set EtcdCluster.Status.Ready to true

Reviewer: I know this may be out of scope. How are `EtcdCluster.Status.Endpoints` consumed by kube-apiserver?

Author: good question, not out of scope, the section `Contract between etcd provider and control plane provider` talks about how control plane providers can consume this field. After that, each control plane provider is responsible for processing the endpoints field and passing it to the kube-apiserver. The default control plane provider in CAPI is KubeadmControlPlane (KCP), and it already has a field for accepting external etcd endpoints and passing them to kube-apiserver. So this default control plane provider will continue using the same logic

- The controller should do healthchecks of the etcd members in a loop at a certain predetermined interval (10s).
- When a member fails healthcheck for a certain amount of time (1m), replace that member with a new member

**Changes needed in docker machine controller**
- Each machine infrastructure provider sets a providerID on every InfraMachine.
- Infrastructure providers like CAPA and CAPV set this provider ID based on the instance metadata.
- CAPD on the other hand first calls `kubectl patch` to add the provider ID on the node. And only then it sets the provider ID value on the DockerMachine resource. But the problem is that in case of an etcd only cluster, the machine is not registered as a Kubernetes node. Cluster provisioning does not progress until this provider ID is set on the Machine. As a solution, CAPD will check if the bootstrap provider is etcdadm, and skip the kubectl patch process and set the provider ID on DockerMachine directly. These changes are required in the [docker machine controller](https://github.com/mrajashree/cluster-api/pull/2/files#diff-1923dd8291a9406e3c144763f526bd9797a2449a030f5768178b8d06d13c795bR307) for CAPD.

**Future work**

> ##### Static Etcd endpoints

Reviewer: Great callout. Does this mean the first iteration would be a set of IPs that may change? Does kube-apiserver need restart?

Author: that's right

    Conditions clusterv1.Conditions `json:"conditions,omitempty"`
    
    DataSecretName *string `json:"dataSecretName,omitempty"`

    Ready bool `json:"ready,omitempty"`
}
```

The following fields will be added/modified in existing CAPI types:
- Add Cluster.Spec.ManagedExternalEtcdRef as:
```go
// ManagedExternalEtcdRef is an optional reference to an etcd provider resource that holds details
// for provisioning an external etcd cluster
// +optional
ManagedExternalEtcdRef *corev1.ObjectReference `json:"managedExternalEtcdRef,omitempty"`
```
- Add Cluster.Status.ManagedExternalEtcdReady field as:
```go
// ManagedExternalEtcdReady indicates external etcd cluster is fully provisioned
// +optional
ManagedExternalEtcdReady bool `json:"managedExternalEtcdReady"`
```
- Make KubeadmConfigSpec.ClusterConfiguration.Etcd.External.Endpoints field mutable
    - The kubeadm validating webhook currently doesn't allow this field to get updated. It should be added as an [allowed path](https://github.com/mrajashree/cluster-api/commit/18c42c47d024ce17cdda39500fc0d6bd67c5aa67#diff-6603d336435f3ee62a50043413f62441b685d5c416e5cadd792ed2a536b2f55fR121) in the webhook.

#### Implementation details

> ##### Contract for pluggable etcd providers

Reviewer: Still not clear about the relationship between "etcd provider", "etcd controller", and "etcdadm".

"etcd controller" is AN "etcd provider" that's built on top of "etcdadm"?

Author: I was not done addressing the comments :)

###### An etcd controller’s main responsibilities are:

* Managing a set of Machines that represent an etcd member.
* Managing the etcd cluster, in terms of scaling up/down/upgrading etcd version, and running healthchecks.
* Populating and updating a field that will contain the etcd cluster endpoints (IP addresses of all etcd members). 
* Creating/managing two Kubernetes Secrets that will contain certs required by the API server to communicate with the etcd cluster
    * One Secret containing the etcd CA cert, by the name `{cluster.Name}-etcd`
    * One Secret containing the API server cert-key pair, by the name `{cluster.Name}-apiserver-etcd-client`

###### Required services:
The etcd provider should install etcd with the version specified by the user for each member.

###### Required fields:
Each etcd provider should add a new API type to manage the etcd cluster lifecycle, and it should have the following required fields:
- Spec:
    - Replicas: Allow only odd numbers, and immutable after creation.
- Status

>     - Ready: To be set to true once all etcd members pass healthcheck (making calls to /health endpoint).

Reviewer: This is `Status` that represents the whole cluster? What is the action when this evaluates to `false`? Or just read-only mode for visibility?

Author: makes sense to not name it "Ready", changed it to "CreationComplete"

    - Endpoint: A comma separated string containing all etcd members endpoints.

> ###### Contract between etcd provider and control plane provider

Reviewer: Where is control plane provider running? Is that for Kubernetes control plane?

Author: I can discuss CAPI specific details offline

  - "@mrajashree"
creation-date: 2021-05-24
last-updated: 2021-05-26
status: provisional
---

# Add support for managed external etcd clusters in CAPI

## Table of Contents

- [Title](#title)
  - [Table of Contents](#table-of-contents)
  - [Glossary](#glossary)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals/Future Work](#non-goalsfuture-work)
  - [Proposal](#proposal)
    - [User Stories](#user-stories)
    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
    - [POC implementation](#poc-implementation)
    - [Questions/Concerns](#questions/concerns-about-implementation)
    - [Design questions](#design-questions)
    - [Security Model](#security-model)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan [optional]](#test-plan-optional)
    - [Graduation Criteria [optional]](#graduation-criteria-optional)
    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
  - [Implementation History](#implementation-history)

## Glossary

### Etcd provider

> A provider that creates and manages an etcd cluster to be used by a single workload cluster for the [external etcd topology](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/#external-etcd-topology).

Author: yeah I agree. But in the section in implementation details that defines the contract for etcd provider, I've used the term 'etcd provider' just like the contract for control plane uses the term 'control plane provider'. And the contract section doesn't mention etcdadm based provider at all, so I wanted to define the term 'etcd provider' to make it a bit clear. Do you think we can leave it in in that case? also we can rename it (couldn't think of a better name)

    
  - External etcd topology decouples the control plane and etcd member. So if a control plane-only node fails, or if there's a memory leak in a component like kube-apiserver, it won't directly impact an etcd member.
    
  - Etcd is resource intensive, so it is safer to have dedicated nodes for etcd, it could use more disk space, higher bandwidth. 
    Having a separate etcd cluster for these reasons could ensure a more resilient HA setup.
    
- Motivation behind using etcdadm as a pluggable etcd provider
  
  - Leveraging existing projects such as etcdadm is one way of bringing up an etcd cluster. One of the goals mentioned in the [etcdadm project roadmap](https://github.com/kubernetes-sigs/etcdadm/blob/master/ROADMAP.md) is to integrate it with CAPI to support the "external etcd" use case.
  
  - Etcd providers can be made pluggable and we can start by adding one that uses etcdadm.

### Goals

- Introduce pluggable etcd providers in Cluster API, starting with an etcdadm based provider. This etcdadm provider will create and manage an etcd cluster.

> - User should be able to create a Kubernetes cluster that uses external etcd topology in a single step.
> - User should only have to create the required CAPI resources, along with the required etcd provider specific resources. That should trigger creation of an etcd cluster, followed by creation of the target workload cluster which uses this etcd cluster. Meaning, the CAPI controllers will deploy the etcd cluster first and then use that as input to the workload cluster, without requiring any user intervention.

Reviewer 2: What about collapsing the two goals into 
```suggestion
- User should be able to create and manage a Kubernetes cluster that uses external etcd topology in a declarative way.
```

- There will be a 1:1 mapping between the external etcd cluster and the workload cluster.
- Define a contract for pluggable etcd providers and define steps that control plane providers should take to use a managed external etcd cluster.
- Support the following etcd cluster management actions: scale up and scale down, etcd member replacement and etcd version upgrades.
- The etcd providers will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.

### Non-Goals

> - The first iteration will use IP addresses/hostnames as etcd cluster endpoints. It can not configure static endpoint(s) till the Load Balancer provider is available.

Reviewer 2: Uhmm, would this trigger a control plane rollout every time the list of address/host names changes?

Author: Yes, it would trigger a control plane rollout in this first iteration since we decided to configure static endpoints through the load balancer provider when it is implemented. Not sure if there is a different way of updating the kube-apiserver, with the etcd endpoints, can't think of one. And for static endpoints can't think of a different approach that won't need changes in the infrastructure providers (apart from the load balancer provider). Ideally the etcd endpoints should only change when etcd machines are replaced, in case of upgrades or if a member is unhealthy and needs to be replaced.

Reviewer 2: What about asking the users to provide a stable address/a dns for the etcd cluster?

Author: yes I have included that in the "Future work" section. In that I've also included how using multiple dns records as static endpoints, one per etcd member is beneficial because through that we can leverage [etcd client's balancer](https://etcd.io/docs/v3.4/learning/design-client/#clientv3-grpc10-balancer-overview).
We discussed in the last CAPI meeting how the contract shouldn't change once we start using static endpoints. Right now the contract section in this KEP specifies that the etcd cluster's endpoint is a list of IP addresses for the first iteration. Later on when we use DNS records as static endpoints, if we configure a dns record per member (to leverage etcd's balancer) we will still have a list of endpoints, so nothing has to change in the contract.

> - API changes such as adding new fields within existing control plane providers, like the KubeadmControlPlane provider. We will utilize the existing external etcd endpoints field from the KubeadmConfigSpec for KCP.

Reviewer 2: This is tricky.
KCP is a pluggable component in CAPI, thus we should not make assumptions it exist in a Cluster, and technically the only assumption we can make on a control plane provider are documented in https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane.html#contracts.

That's means "We will utilize the existing external etcd endpoints field from the KubeadmConfigSpec for KCP." can't be used or the scope of this proposal must be changed so we are targeting only cluster with KCP (but this is really limiting)

Author: the intention was to say that we won't add any new fields to the existing control plane provider API types specifically for external etcd clusters. And I wanted to include KCP only as an example. I have rephrased this in an attempt to make that clear. Does that look better now or should I remove this, because I agree that changing the scope of this proposal to target KCP only is limiting.

Author: Modified the proposal goals to make this external etcd provider work only with KCP for now. Added the proposed change in control plane contract in the non-goals/future work section so we can treat it as a separate proposal

> - An etcd cluster created for one workload cluster should not be used as etcd for another workload cluster. This proposal does not intend to add 1:many mapping for etcd to workload clusters.

Reviewer 2: This non goal is the same of the 1:1 goal above. What about removing?
